src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+                } catch (ItemNotFound $e) {
+                    return;
+                }
+
+                if ($version > $current) {
+                    return;
+                }
+
+                if ($version < $current) {
+                    throw new UnexpectedVersionNumber($id, $version, $current);
+                }
+
+                if (1 === $current->toInt()) {
+                    $this->deleteItem($id);
+
+                    return;
+                }
+
+                $this->deleteVersion($id, $current);
+            }
+        );
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (null === $version) {
+            $sql = 'SELECT version.id, version.version, version.content, version.hash'
+                ." FROM {$this->tableItems} item LEFT JOIN {$this->tableVersions} version"
+                .' ON item.id = version.id AND item.version = version.version'
+                .' WHERE item.id = :id';
+            $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id]);
+
+            $row = $stmt->fetch(FetchMode::ASSOCIATIVE);
+
+            if (false === $row) {
+                throw new ItemNotFound($id);
+            }
+
+            return $this->rowToItemVersion($row);
+        }
+
+        $sql = 'SELECT item.id, version.version, version.content, version.hash'
+            ." FROM {$this->tableItems} item LEFT JOIN {$this->tableVersions} version"
+            .' ON item.id = version.id AND version.version = :version WHERE item.id = :id';
+        $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id, 'version' => $version->toInt()]);
+
+        $row = $stmt->fetch(FetchMode::ASSOCIATIVE);
+
+        if (!$row) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $row['version']) {
+            throw new VersionNotFound($id, $version);
+        }
+
+        return $this->rowToItemVersion($row);
+    }
+
+    public function list(int $limit = 10, ?string $cursor = null) : ItemListPage
+    {
+        if (null !== $cursor && ((string) (int) $cursor) !== $cursor) {
+            return new ItemListPage([], null);
+        }
+
+        $query = $this->connection->createQueryBuilder();
+
+        $query
+            ->addSelect('item.id')
+            ->addSelect('item.sequence')
+            ->from($this->tableItems, 'item')
+            ->orderBy('item.sequence')
+            ->setMaxResults($limit + 1);
+
+        if ($cursor) {
+            $query
+                ->where($query->expr()->gte('item.sequence', ':cursor'))
+                ->setParameter('cursor', (int) $cursor);
+        }
+
+        $results = $query->execute();
+        $ids = $results->fetchAll(FetchMode::ASSOCIATIVE);
+
+        if (empty($ids)) {
+            return new ItemListPage([], null);
+        }
+
+        if (count($ids) > $limit) {
+            $newCursor = (string) array_pop($ids)['sequence'];
```

Comment:
You can implement page-based paging using cursors (it would be the page number in that case), depends if you're willing to accept those alignment problems.

I'd leave for now, but it might be possible to add in a paging strategy option.

Reply:
Overkill to have various pagination strategy options in a `article title|page` milestone